common/flatpak-dir.c

```diff
@@ -7718,6 +7817,74 @@ flatpak_dir_create_system_child_repo (FlatpakDir   *self,
   return flatpak_dir_create_child_repo (self, cache_dir, file_lock, optional_commit, error);
 }
 
+static gboolean
+flatpak_dir_setup_revokefs_fuse_mount (FlatpakDir    *self,
+                                       const gchar   *ref,
+                                       const gchar   *installation,
+                                       gchar        **src_dir,
+                                       gchar        **mnt_dir,
+                                       GCancellable  *cancellable)
+{
+  g_autoptr (GError) local_error = NULL;
+  g_autofree gchar *src_dir_tmp = NULL;
+  g_autofree gchar *mnt_dir_tmp = NULL;
+  gint socket = -1;
+  gboolean res = FALSE;
+
+  if (!flatpak_dir_system_helper_call_get_revokefs_fd (self,
+                                                       FLATPAK_HELPER_GET_REVOKEFS_FD_FLAGS_NONE,
+                                                       installation ? installation : "",
+                                                       &socket,
+                                                       &src_dir_tmp,
+                                                       cancellable,
+                                                       &local_error))
+    {
+      if (g_error_matches (local_error, G_DBUS_ERROR, G_DBUS_ERROR_NOT_SUPPORTED))
+        g_debug ("revokefs-fuse not supported on your installation: %s", local_error->message);
+      else
+        g_warning ("Failed to get revokefs-fuse socket from system-helper: %s", local_error->message);
+
+      goto out;
+    }
+  else
+    {
+      g_autoptr(GSubprocess) revokefs_fuse = NULL;
+      g_autoptr(GSubprocessLauncher) launcher = NULL;
+      g_autofree gchar *client_uid = NULL;
+
+      mnt_dir_tmp = flatpak_dir_revokefs_fuse_create_mountpoint (ref, &local_error);
+      if (mnt_dir_tmp == NULL)
+        {
+          g_warning ("Failed to create a mountpoint for revokefs-fuse: %s", local_error->message);
+          close (socket);
+          goto out;
+        }
+
+      client_uid = g_strdup_printf ("uid=%d", getuid ());
+      launcher = g_subprocess_launcher_new (G_SUBPROCESS_FLAGS_NONE);
+      g_subprocess_launcher_take_fd (launcher, socket, 3);
+      revokefs_fuse = g_subprocess_launcher_spawn (launcher,
+                                                   &local_error,
+                                                   "revokefs-fuse", "-o", client_uid, "--socket=3",
+                                                   src_dir_tmp, mnt_dir_tmp, NULL);
+      if (revokefs_fuse == NULL ||
+          !g_subprocess_wait_check (revokefs_fuse, NULL, &local_error))
+        {
+          g_warning ("Error spawning revokefs-fuse: %s", local_error->message);
+          close (socket);
+          goto out;
+        }
+    }
+
+  res = TRUE;
+
+out:
+  *mnt_dir = g_steal_pointer (&mnt_dir_tmp);
```

Comment:
In case of an error we should not return these, that will likely cause a leak in the parent.

Reply:
(It doesn't actually leak in the current caller, but still it could.)

Reply:
Ah, I see. 
There was a earlier [comment](https://github.com/flatpak/flatpak/pull/2657#discussion_r266973414) for the same function  hence, I re-wrote it this way.
> Only set this at the end, or we will leak it on error. Also, better to us g_steal_pointer() to avoid duplicating it.

The issue here is that the `src_dir_tmp` can be populated but it fails for create mountpoint (hence, `mnt_dir_tmp` is NULL). Hence, cleanup should be done accordingly (i.e. if `src_dir != NULL` -> Call `CancelPull` dbus). See next commit where I added CancelPull and `flatpak_dir_unmount_and_cancel_pull` function. 

Will think more on it later.

Reply:
I see, that makes sense. However, it is standard to only return out params on success and clean up internally on error. So we should add some comment to this function about this nonstandard behaviour and cleanup requirements.

Reply:
Added a comment explaining the situation.